### DISCUSSION TITLE
DynamicSceneBuilder error: "the given key does not support hashing"

### DISCUSSION THREAD

**Discussion by Liam19** [bevyengine/bevy#13029] [Q&A] [answered]
Im trying to build a simple DynamicScene using the DynamicSceneBuilder in my save_game system:

```
#[derive(Clone, Deserialize, Resource, Reflect, Default, Debug, Hash)]
#[reflect(Resource)]
pub struct ActorNames {
    pub male_names: Vec<String>,
    pub female_names: Vec<String>,
    pub last_names: Vec<String>,
}

fn save_game(world: &mut World) {
    let builder: DynamicSceneBuilder = DynamicSceneBuilder::from_world(world);
    //Fails
    let builder_filtered: DynamicSceneBuilder = builder.extract_resources().allow_resource::<Names>();
}
```
But i keep getting the error: "...bevy_reflect-0.13.2\src\map.rs:292:53: the given key does not support hashing".

I've tried using multiple different resources but they all give the same error. They are also all registered in the type registry. What am i doing wrong? Is there a better way to extract Resources into a DynamicScene?

**Comment by MrGVSV** [comment, marked as answer]
Note that `extract_***` methods are invoked immediately so your `allow_resource` call has no effect.

This makes me wonder if it's some Bevy-owned resource that's actually causing the error. Do you still face the same problem when you put the filter call before the extraction?

**Reply by Liam19** [reply]
Putting the filter call before the extraction works for the ActorNames resource but now i see another problem. I have this struct:

```
#[derive(Default, Copy, Clone, PartialEq, Eq, Hash, Reflect, Debug)]
pub struct Item(pub u32);

impl<'de> Deserialize<'de> for Item {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let item_name = String::deserialize(deserializer)?;

        Ok(string_to_item(&item_name))
    }
}
```
Whenever a resource has a `HashMap<Item, _>` field, i get the error described above. 
I had a look at the stack backtrace and found this: 
bevy_reflect::impls::std::impl$49::clone_dynamic<item_module::item::Item, ...
Seems like there might be a problem with `clone_dynamic` for the Item struct?


EDIT:
I added `#[reflect(Hash)]` to `Item` and it now works perfectly! Thanks for your help.
